alchemiscale_fah/compute/index.py

```diff
@@ -0,0 +1,298 @@
+"""
+:mod:`alchemiscale_fah.compute.index` --- compute service index interface
+=========================================================================
+
+"""
+
+import os
+import plyvel
+import json
+import shutil
+from typing import List, Tuple, Optional
+from pathlib import Path
+
+from gufe.tokenization import GufeKey, JSON_HANDLER, GufeTokenizable
+from gufe.protocols.protocoldag import ProtocolDAG
+from gufe.protocols.protocolunit import ProtocolUnitResult
+from alchemiscale.models import ScopedKey
+from alchemiscale.keyedchain import KeyedChain
+
+from .models import FahProject, FahRun, FahClone
+
+
+class FahComputeServiceIndex:
+    """Persistent index interface for FahComputeServices"""
+
+    def __init__(self, index_dir: os.PathLike, obj_store: Optional[os.PathLike] = None):
+        self.index_dir = Path(index_dir).absolute()
+        self.index_dir.mkdir(parents=True, exist_ok=True)
+
+        self.db = plyvel.DB(str(self.index_dir), create_if_missing=True)
+
+        # self.projects = self.db.prefixed_db(b"projects/")
+        # self.runs = self.db.prefixed_db(b"runs/")
+        # self.clones = self.db.prefixed_db(b"clones/")
+
+        # self.transformations = self.db.prefixed_db(b"transformations/")
+        # self.tasks = self.db.prefixed_db(b"tasks/")
+
+        self.obj_store = Path(obj_store).absolute()
+        self.obj_store.mkdir(parents=True, exist_ok=True)
+
+    def set_project(self, project_id: str, fah_project: FahProject):
+        """Set the metadata for the given PROJECT."""
+        key = f"projects/{project_id}".encode("utf-8")
+        value = fah_project.json().encode("utf-8")
+
+        self.db.put(key, value)
+
+    def get_project(self, project_id: str) -> FahProject:
+        """Get metadata for the given PROJECT."""
+        key = f"projects/{project_id}".encode("utf-8")
+        value = self.db.get(key)
+
+        if value is not None:
+            value = FahProject.parse_raw(value.decode("utf-8"))
+
+        return value
+
+    def get_project_run_next(self, project_id: str) -> FahProject:
+        """Get next available RUN id for the given PROJECT."""
+        prefix = f"runs/{project_id}-".encode("utf-8")
+        run_ids = sorted(
+            [
+                int(key.decode("utf-8").split("-")[-1])
+                for key in self.db.iterator(prefix=prefix, include_value=False)
+            ]
+        )
+
+        if run_ids:
+            return str(run_ids[-1] + 1)
+        else:
+            return "0"
+
+    def set_run(self, project_id: str, run_id: str, fah_run: FahRun):
```

Comment:
Fair point, not really sure of the best approach here, since we have to convert everything into bytestring representations for LevelDB anyway. Should we just switch to using integers for all `project_id`, `run_id`, and `clone_id`s throughout the codebase so that our models might catch this?

Reply:
Updated all index methods to take `int`s for `project_id`, `run_id`, and `clone_id`, and return the same. We also now store these entities as `int`s in the `FahProject`, `FahRun`, and `FahClone` models.